flink-ml-core/src/main/java/org/apache/flink/ml/common/broadcast/BroadcastUtils.java

```diff
@@ -152,41 +155,52 @@ private static <OUT> DataStream<OUT> cacheBroadcastVariables(
-     * @param <OUT> output type of the result stream.
-     * @return the result stream by applying user-defined logic on the input list.
+     * @param env Execution environment.
+     * @param inputList Non-broadcast input list.
+     * @param broadcastStreamNames Names of the broadcast data streams.
+     * @param graphBuilder User-defined logic.
+     * @param <OUT> Output type of the result stream.
+     * @return The result stream by applying user-defined logic on the input list.
      */
     private static <OUT> DataStream<OUT> getResultStream(
             StreamExecutionEnvironment env,
             List<DataStream<?>> inputList,
             String[] broadcastStreamNames,
             Function<List<DataStream<?>>, DataStream<OUT>> graphBuilder) {
-        TypeInformation<?>[] inTypes = new TypeInformation[inputList.size()];
-        for (int i = 0; i < inputList.size(); i++) {
-            inTypes[i] = inputList.get(i).getType();
-        }
-        // do not block all non-broadcast input edges by default.
-        boolean[] isBlocked = new boolean[inputList.size()];
-        Arrays.fill(isBlocked, false);
+
+        // Executes the graph builder and gets real non-broadcast inputs.
         DraftExecutionEnvironment draftEnv =
-                new DraftExecutionEnvironment(
-                        env, new BroadcastWrapper<>(broadcastStreamNames, inTypes, isBlocked));
+                new DraftExecutionEnvironment(env, new DefaultWrapper<>());
 
         List<DataStream<?>> draftSources = new ArrayList<>();
         for (DataStream<?> dataStream : inputList) {
             draftSources.add(draftEnv.addDraftSource(dataStream, dataStream.getType()));
         }
         DataStream<OUT> draftOutStream = graphBuilder.apply(draftSources);
-        Preconditions.checkState(
-                draftEnv.getStreamGraph(false).getStreamNodes().size() == 1 + inputList.size(),
-                "cannot add more than one operator in withBroadcastStream's lambda function.");
-        draftEnv.copyToActualEnvironment();
-        return draftEnv.getActualStream(draftOutStream.getId());
+
+        List<Transformation<?>> realNonBroadcastInputs =
+                draftOutStream.getTransformation().getInputs();
+        TypeInformation<?>[] inTypes = new TypeInformation[realNonBroadcastInputs.size()];
+        for (int i = 0; i < realNonBroadcastInputs.size(); i++) {
+            inTypes[i] = realNonBroadcastInputs.get(i).getOutputType();
+        }
+        // Does not block all non-broadcast input edges by default.
+        boolean[] isBlocked = new boolean[realNonBroadcastInputs.size()];
+        Arrays.fill(isBlocked, false);
+
+        // Executes the graph builder and adds the operators to real execution environment.
+        DraftExecutionEnvironment draftEnv2 =
+                new DraftExecutionEnvironment(
+                        env, new BroadcastWrapper<>(broadcastStreamNames, inTypes, isBlocked));
```

Comment:
Is it possible for `AbstractBroadcastWrapperOperator` to retrieve inTypes from `wrappedOperator` directly instead of getting this information from the constructor? 

If we can do this, we can make the code more readable by only invoking `graphBuilder` only once here.

Maybe add a TODO here if it is not straightforward to make this change in this PR.

Reply:
Good catch. I have updated the implementation as you suggested. Please checkout the latest commit.